File: packages/sdk/src/types/pricing.ts
// ---------------------------------------------------------------------------
// Core pricing types for the Theatrical SDK
// ---------------------------------------------------------------------------

/** Cinema price format — standard, IMAX, 4DX, ScreenX, Dolby Atmos, etc. */
export type SessionFormat = 'standard' | 'imax' | '4dx' | 'screenx' | 'dolby' | 'vmax' | 'gold-class';

/** Day-part classification affecting matinee/peak pricing. */
export type DayPart = 'matinee' | 'afternoon' | 'peak' | 'late';

/** Ticket type grouping — matches Vista OCAPI ticket type categories. */
export type TicketCategory = 'adult' | 'child' | 'senior' | 'student' | 'family' | 'concession' | 'loyalty-member';

// ---------------------------------------------------------------------------
// Ticket types
// ---------------------------------------------------------------------------

export interface TicketType {
  id: string;
  name: string;
  description?: string;
  /** Base price in minor currency units (e.g., cents). */
  price: number;
  currency: string;
  category: TicketCategory;
  isDefault: boolean;
  requiresLoyalty: boolean;
  /** Minimum age for this ticket type (e.g. senior >= 65). */
  minimumAge?: number;
  /** Maximum age (e.g. child <= 14). */
  maximumAge?: number;
  /** Whether this ticket type is currently purchasable. */
  isAvailable: boolean;
}

export interface TicketTypeFilter {
  sessionId: string;
  /** Filter by category (e.g., only loyalty-eligible types). */
  category?: TicketCategory;
  /** Exclude unavailable ticket types. Defaults to true. */
  availableOnly?: boolean;
}

// ---------------------------------------------------------------------------
// Tax configuration
// ---------------------------------------------------------------------------

/**
 * Regional tax configuration for price display.
 *
 * NZ cinema operators present tax-inclusive prices; Australian operators
 * sometimes break GST out separately in B2B contexts.
 */
export interface TaxConfig {
  /** ISO 4217 currency code. */
  currency: string;
  /** Tax rate as a decimal fraction (e.g. 0.15 for 15% NZ GST). */
  rate: number;
  /** Label shown on receipts (e.g. "GST", "VAT"). */
  label: string;
  /**
   * When true, all prices are tax-inclusive (consumer-facing display).
   * When false, tax is added on top (B2B / booking-system context).
   */
  inclusive: boolean;
}

// ---------------------------------------------------------------------------
// Discounts & surcharges
// ---------------------------------------------------------------------------

export type DiscountSource = 'loyalty-tier' | 'coupon' | 'promo' | 'member' | 'employee' | 'group';

export interface Discount {
  id: string;
  source: DiscountSource;
  /** Human-readable label (e.g. "Gold Member Discount"). */
  label: string;
  /** Discount amount in minor units. */
  amount: number;
  /** Percentage discount, 0–100. Informational only when `amount` is authoritative. */
  percentage?: number;
  /** Coupon code that triggered this discount, if applicable. */
  couponCode?: string;
}

export type SurchargeReason = 'format' | 'seat-type' | 'booking-fee' | 'service-fee' | 'peak-surcharge';

export interface Surcharge {
  id: string;
  reason: SurchargeReason;
  /** Human-readable label (e.g. "IMAX Surcharge"). */
  label: string;
  /** Surcharge amount in minor units. */
  amount: number;
}

// ---------------------------------------------------------------------------
// Price breakdown & calculation
// ---------------------------------------------------------------------------

/**
 * Itemised price breakdown for a single ticket or a quantity of tickets.
 *
 * All amounts are in minor units of `currency` (e.g. cents for NZD).
 */
export interface PriceBreakdown {
  /** Base price per ticket before adjustments. */
  basePrice: number;
  /** Ordered list of discounts applied. */
  discounts: Discount[];
  /** Ordered list of surcharges applied. */
  surcharges: Surcharge[];
  /** Tax amount. Zero when `taxConfig.inclusive` is true. */
  taxAmount: number;
  /** Sum of all discounts. */
  totalDiscount: number;
  /** Sum of all surcharges. */
  totalSurcharge: number;
  /** Per-ticket price after all adjustments. */
  pricePerTicket: number;
  /** `pricePerTicket * quantity`. */
  totalPrice: number;
  quantity: number;
  currency: string;
  taxConfig: TaxConfig;
}

/**
 * Full price calculation response from the Vista pricing engine.
 *
 * Returned by `PricingResource.calculate()`. The `breakdown` field gives
 * full itemisation; legacy integrations can use the flat fields.
 */
export interface PriceCalculation {
  sessionId: string;
  ticketTypeId: string;
  /** Total price, minor units. */
  totalPrice: number;
  currency: string;
  /** Whether returned price includes tax. */
  taxInclusive: boolean;
  /** Detailed, itemised breakdown. */
  breakdown: PriceBreakdown;
  /** ISO 8601 timestamp — price is valid until this time. */
  validUntil: string;
}

// ---------------------------------------------------------------------------
// Coupon application
// ---------------------------------------------------------------------------

export interface ApplyCouponsInput {
  sessionId: string;
  ticketTypeId: string;
  quantity: number;
  couponCodes: string[];
  /** Loyalty member ID for tier-based discount resolution. */
  membershipId?: string;
}

export interface CouponApplicationResult {
  applied: Discount[];
  rejected: Array<{ code: string; reason: string }>;
  updatedBreakdown: PriceBreakdown;
}

File: packages/sdk-py/theatrical/types/pricing.py
"""Pricing types."""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class SessionPricingFormat(str, Enum):
    STANDARD = "standard"
    IMAX = "imax"
    FOUR_DX = "4dx"
    SCREENX = "screenx"
    DOLBY = "dolby"
    VMAX = "vmax"
    GOLD_CLASS = "gold-class"


class DayPart(str, Enum):
    MATINEE = "matinee"
    AFTERNOON = "afternoon"
    PEAK = "peak"
    LATE = "late"


class TicketCategory(str, Enum):
    ADULT = "adult"
    CHILD = "child"
    SENIOR = "senior"
    STUDENT = "student"
    FAMILY = "family"
    CONCESSION = "concession"
    LOYALTY_MEMBER = "loyalty-member"


class TicketType(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    price: float
    currency: str
    category: TicketCategory
    is_default: bool
    requires_loyalty: bool
    minimum_age: Optional[int] = None
    maximum_age: Optional[int] = None
    is_available: bool


class TicketTypeFilter(BaseModel):
    session_id: str
    category: Optional[TicketCategory] = None
    available_only: Optional[bool] = None


class TaxConfig(BaseModel):
    currency: str
    rate: float
    label: str
    inclusive: bool


class DiscountSource(str, Enum):
    LOYALTY_TIER = "loyalty-tier"
    COUPON = "coupon"
    PROMO = "promo"
    MEMBER = "member"
    EMPLOYEE = "employee"
    GROUP = "group"


class Discount(BaseModel):
    id: str
    source: DiscountSource
    label: str
    amount: float
    percentage: Optional[float] = None
    coupon_code: Optional[str] = None


class SurchargeReason(str, Enum):
    FORMAT = "format"
    SEAT_TYPE = "seat-type"
    BOOKING_FEE = "booking-fee"
    SERVICE_FEE = "service-fee"
    PEAK_SURCHARGE = "peak-surcharge"


class Surcharge(BaseModel):
    id: str
    reason: SurchargeReason
    label: str
    amount: float


class PriceBreakdown(BaseModel):
    base_price: float
    discounts: list[Discount]
    surcharges: list[Surcharge]
    tax_amount: float
    total_discount: float
    total_surcharge: float
    price_per_ticket: float
    total_price: float
    quantity: int
    currency: str
    tax_config: TaxConfig


class PriceCalculation(BaseModel):
    session_id: str
    ticket_type_id: str
    total_price: float
    currency: str
    tax_inclusive: bool
    breakdown: PriceBreakdown
    valid_until: str


class ApplyCouponsInput(BaseModel):
    session_id: str
    ticket_type_id: str
    quantity: int
    coupon_codes: list[str]
    membership_id: Optional[str] = None


class RejectedCoupon(BaseModel):
    code: str
    reason: str


class CouponApplicationResult(BaseModel):
    applied: list[Discount]
    rejected: list[RejectedCoupon]
    updated_breakdown: PriceBreakdown
